File: kotlin/src/main/kotlin/com/cardinal/modules/Codification.kt
package com.cardinal.modules

import com.icure.cardinal.sdk.filters.ServiceFilters
import com.icure.cardinal.sdk.model.Code
import com.icure.cardinal.sdk.model.DecryptedContact
import com.icure.cardinal.sdk.model.DecryptedPatient
import com.icure.cardinal.sdk.model.base.CodeStub
import com.icure.cardinal.sdk.model.embed.DecryptedContent
import com.icure.cardinal.sdk.model.embed.DecryptedService
import com.icure.cardinal.sdk.model.embed.Measure
import com.cardinal.utils.prettyPrint
import com.icure.cardinal.sdk.CardinalSdk
import com.icure.cardinal.sdk.filters.CodeFilters
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import kotlin.random.Random

suspend fun manageCodification(sdk: CardinalSdk) {
	try {
		val existing = sdk.code.getCodes(
			listOf("INTERNAL|ANALYSIS|1", "SNOMED|45007003|1", "SNOMED|38341003|1", "SNOMED|2004005|1")
		)

		if(existing.isEmpty()) {
			val internalCode = sdk.code.createCode(Code(
				id = "INTERNAL|ANALYSIS|1",
				type = "INTERNAL",
				code = "ANALYSIS",
				version = "1",
				label = mapOf("en" to "Internal analysis code")
			))
			prettyPrint(internalCode)
			sdk.code.createCodes(listOf(
				Code(
					id = "SNOMED|45007003|1",
					type = "SNOMED",
					code = "45007003",
					version = "1",
					label = mapOf("en" to "Low blood pressure")
				),
				Code(
					id = "SNOMED|38341003|1",
					type = "SNOMED",
					code = "38341003",
					version = "1",
					label = mapOf("en" to "High blood pressure")
				),
				Code(
					id = "SNOMED|2004005|1",
					type = "SNOMED",
					code = "2004005",
					version = "1",
					label = mapOf("en" to "Normal blood pressure")
				)
			))
		}

		val codeIterator = sdk.code.filterCodesBy(
			CodeFilters.byLanguageTypeLabelRegion(
				language = "en",
				label = "blood",
				type = "SNOMED"
			)
		)

		var selectedCode: Code? = null
		while (codeIterator.hasNext() && selectedCode == null) {
			val code = codeIterator.next(1).first()
			prettyPrint(code)
			print("Use this code? [y/N]: ")
			val use = readln().trim().lowercase() == "y"
			if (use) {
				selectedCode = code
			}
		}

		if (selectedCode == null) {
			println("No code was selected")
			return
		}

		val patient = sdk.patient.createPatient(
			DecryptedPatient(
				id = UUID.randomUUID().toString(),
				firstName = "Annabelle",
				lastName = "Hall",
			).let { sdk.patient.withEncryptionMetadata(it) }
		)

		val formatter = DateTimeFormatter.ofPattern("yyyyMMddHHmmss")
		val contact = DecryptedContact(
			id = UUID.randomUUID().toString(),
			descr = "Blood pressure measurement",
			openingDate = LocalDateTime.now().format(formatter).toLong(),
			services = setOf(
				DecryptedService(
					id = UUID.randomUUID().toString(),
					label = "Blood pressure",
					content = mapOf(
						"en" to DecryptedContent(
							measureValue = Measure(
								value = Random.nextInt(80, 120).toDouble(),
								unit = "mmHg"
							)
						)
					),
					tags = setOf(
						CodeStub(
							id = selectedCode.id,
							type = selectedCode.type,
							code = selectedCode.code,
							version = selectedCode.version
						)
					)
				)
			)
		)
		val createdContact = sdk.contact.createContact(
			sdk.contact.withEncryptionMetadata(contact, patient)
		)
		prettyPrint(createdContact)
		prettyPrint(createdContact.services.first())

		val serviceIterator = sdk.contact.filterServicesBy(
			ServiceFilters.byTagAndValueDateForSelf(
				tagType = selectedCode.type!!,
				tagCode = selectedCode.code
			)
		)

		println("Result of searching Services by code: ${selectedCode.id}")
		while (serviceIterator.hasNext()) {
			val service = serviceIterator.next(1).first()
			prettyPrint(service)
		}
	} catch (e: Exception) {
		println("Something went wrong: ${e.message}")
	}
}

File: python/src/modules/codification.py
import random
import uuid
from datetime import datetime
from cardinal_sdk import CardinalSdk
from cardinal_sdk.filters import CodeFilters
from cardinal_sdk.filters import ServiceFilters
from cardinal_sdk.model import Code, CodeStub, DecryptedContact, DecryptedContent, DecryptedPatient, DecryptedService, \
	Measure
from utils import pretty_print_code, pretty_print_contact, pretty_print_service


def manage_codifications(sdk: CardinalSdk):
	try:
		existing = sdk.code.get_codes_blocking(
			["INTERNAL|ANALYSIS|1", "SNOMED|45007003|1", "SNOMED|38341003|1", "SNOMED|2004005|1"]
		)

		if len(existing) == 0:
			internal_code = sdk.code.create_code_blocking(
				Code(
					id="INTERNAL|ANALYSIS|1",
					type="INTERNAL",
					code="ANALYSIS",
					version="1",
					label={"en": "Internal analysis code"}
				)
			)
			pretty_print_code(internal_code)
			sdk.code.create_codes_blocking(
				Code(
					id="SNOMED|45007003|1",
					type="SNOMED",
					code="45007003",
					version="1",
					label={"en": "Low blood pressure"}
				),
				Code(
					id="SNOMED|38341003|1",
					type="SNOMED",
					code="38341003",
					version="1",
					label={"en": "High blood pressure"}
				),
				Code(
					id="SNOMED|2004005|1",
					type="SNOMED",
					code="2004005",
					version="1",
					label={"en": "Normal blood pressure"}
				)
			)

		code_iterator = sdk.code.filter_codes_by_blocking(
			CodeFilters.by_language_type_label_region(
				language="en",
				label="blood",
				type="SNOMED"
			)
		)

		selected_code = None
		while selected_code is None and code_iterator.has_next_blocking():
			code = code_iterator.next_blocking(1)[0]
			pretty_print_code(code)
			choice = input("Use this code [y/N]: ")
			if choice.lower() == "y":
				selected_code = code
				break

		patient = sdk.patient.create_patient_blocking(
			sdk.patient.with_encryption_metadata_blocking(
				DecryptedPatient(
					id=str(uuid.uuid4()),
					first_name="Annabelle",
					last_name="Hall"
				)
			)
		)

		contact = DecryptedContact(
			id=str(uuid.uuid4()),
			descr="Blood pressure measurement",
			opening_date=int(datetime.now().strftime("%Y%m%d%H%M%S")),
			services=[
				DecryptedService(
					id=str(uuid.uuid4()),
					label="Blood pressure",
					content={
						"en": DecryptedContent(
							measure_value=Measure(
								value=float(random.randint(80, 120)),
								unit="mmHg"
							)
						)
					},
					tags=[
						CodeStub(
							id=selected_code.id,
							type=selected_code.type,
							code=selected_code.code,
							version=selected_code.version
						)
					]
				)
			]
		)
		created_contact = sdk.contact.create_contact_blocking(
			sdk.contact.with_encryption_metadata_blocking(contact, patient)
		)
		pretty_print_contact(created_contact)
		pretty_print_service(created_contact.services[0])

		service_iterator = sdk.contact.filter_services_by_blocking(
			ServiceFilters.by_tag_and_value_date_for_self(
				tag_type=selected_code.type,
				tag_code=selected_code.code
			)
		)
		print(f"Result of searching Services by code: {selected_code.id}")
		while service_iterator.has_next_blocking():
			service = service_iterator.next_blocking(1)[0]
			pretty_print_service(service)

	except Exception as e:
		print(f"Something went wrong: {e}")
